### PR TITLE
awscli: sync with botocore

### LINKED ISSUE #310105: AWSCLI and botocore out of sync
### Describe the bug
The version of AWSCLI in master is 1.32.58. This version requires botocore==1.34.58, however the version of botocore in master is 1.34.87
https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/botocore/default.nix#L17

AWSCli botocore pin: https://github.com/aws/aws-cli/blob/1.32.58/setup.py#L27

### Screenshots
```
> Checking runtime dependencies for awscli-1.32.58-py3-none-any.whl
       >   - botocore==1.34.58 not satisfied by version 1.34.87
```

### Additional context
This started happening after changing my `pkgs.nixVersion.unstable` to `pkgs.nixVersion.latest`

### Notify maintainers

@anthonyroussel

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
 - system: `"x86_64-linux"`
 - host os: `Linux 6.8.7, NixOS, 24.05 (Uakari), 24.05.20240427.58a1abd`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.22.0`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/store/801l7gvdz7yaibhjsxqx82sc7zkakjbq-source`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by natsukium**
## Description of changes

As noted in the comment, the versions of awscli and botocore need to be synchronized.

closes https://github.com/NixOS/nixpkgs/issues/310105

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by natsukium** [comment]
Result of `nixpkgs-review pr 310068` at 0a097c73 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages built successfully:</summary>
  <ul>
    <li>awscli</li>
    <li>bash-my-aws</li>
    <li>git-remote-codecommit</li>
    <li>iceshelf</li>
    <li>nimbo</li>
    <li>pacu</li>
  </ul>
</details>

**Review by ghost — COMMENTED** [review]
pyproject.toml already specifies the botocore version, perhaps just adding an assert and a comment in pythonRelaxDeps would suffice? (and thanks for fixing!)

<details><summary>assert.diff</summary>

```diff
diff --git a/pkgs/tools/admin/awscli/default.nix b/pkgs/tools/admin/awscli/default.nix
index f229c8999360..e19577203320 100644
--- a/pkgs/tools/admin/awscli/default.nix
+++ b/pkgs/tools/admin/awscli/default.nix
@@ -8,7 +8,8 @@
 , awscli
 }:
 
-python3.pkgs.buildPythonApplication rec {
+let
+clipkg = python3.pkgs.buildPythonApplication rec {
   pname = "awscli";
   version = "1.32.58"; # N.B: if you change this, change botocore and boto3 to a matching version too
   pyproject = true;
@@ -23,7 +24,7 @@ python3.pkgs.buildPythonApplication rec {
   ];
 
   pythonRelaxDeps = [
-    "botocore"
+    # botocore must not be relaxed
     "colorama"
     "docutils"
     "rsa"
@@ -86,4 +87,7 @@ python3.pkgs.buildPythonApplication rec {
     mainProgram = "aws";
     maintainers = with maintainers; [ anthonyroussel ];
   };
-}
+};
+in
+assert clipkg ? pythonRelaxDeps -> !(lib.elem "botocore" clipkg.pythonRelaxDeps);
+clipkg
```

**Comment by natsukium** [comment]
It sounds good to introduce an assert.
Since botocore and awscli(v1) have been using the same versioning for a long time, I think it is better to set the version dynamically.

**Comment by natsukium** [comment]
I applied the suggested patch. Thanks.

File: pkgs/tools/admin/awscli/default.nix [hunk 1]
```diff
@@ -8,82 +8,89 @@
 , awscli
 }:
 
-python3.pkgs.buildPythonApplication rec {
-  pname = "awscli";
-  version = "1.32.101"; # N.B: if you change this, change botocore and boto3 to a matching version too
-  pyproject = true;
+let
+  botocoreVersion = python3.pkgs.botocore.version;
+  versionMinor = toString (lib.toInt (lib.versions.minor botocoreVersion) - 2);
```
Comment (annaleeleaves) [inline pkgs/tools/admin/awscli/default.nix:13]:
```suggestion
  # awscli.version should be pinned to 2 minor versions less than the botocoreVersion
  versionMinor = toString (lib.toInt (lib.versions.minor botocoreVersion) - 2);
```

File: pkgs/tools/admin/awscli/default.nix [hunk 2]
```diff
@@ -8,82 +8,89 @@
...
+    version = "1.${versionMinor}.${versionPatch}"; # N.B: if you change this, change botocore and boto3 to a matching version too
+    pyproject = true;
 
-  src = fetchPypi {
-    inherit pname version;
-    hash = "sha256-M4KzuQ+srPfk85P5QeQfVMorSKWhmNhdX4AY0AeRDTQ=";
-  };
+    src = fetchPypi {
+      inherit pname version;
+      hash = "sha256-GPcXkl2H0XNaeqt2/qD5+KvW23dRB0X+zLWo9hLigQM=";
```
Comment (annaleeleaves) [inline pkgs/tools/admin/awscli/default.nix:22]: the version will be bumped when `botocore` is updated and this hash will be invalidated. maybe botocore update script can also update the hash for awscli?

**Review by ghost — COMMENTED** [review]
Looks good, though, won't the hash be a problem when botocore is updated?

File: pkgs/tools/admin/awscli/default.nix (hunk 2 above)
Comment (natsukium) [reply pkgs/tools/admin/awscli/default.nix:23]: Yes, that's why I updated the updateScript.
I will update botocore's updateScript as well.

**Pull request merged**